### PR TITLE
[next] Add support for `home` and `global` as edge function regions

### PR DESCRIPTION
**Pull request by timneutkens**
Adds `home` and `global` as supported values for `regions` when coming from Next.js. We'll port these to the build output API at a later time.

Related Next.js changes: https://github.com/vercel/next.js/pull/48959

### CODE REVIEW

File: packages/next/src/utils.ts [hunk 1]
```diff
@@ -2377,6 +2379,45 @@ interface EdgeFunctionMatcher {
...
+let vercelFunctionRegions: string[] | undefined;
+if (vercelFunctionRegionsVar) {
+  vercelFunctionRegions = vercelFunctionRegionsVar.split(',');
+}
+
+/**
+ * Normalizes the regions config that comes from the Next.js edge functions manifest.
+ * Ensures that config like `home` and `global` are converted to the corresponding Vercel region config.
+ * In the future we'll want to make `home` and `global` part of the Build Output API.
+ * - `home` refers to the regions set in vercel.json or on the Vercel dashboard project config.
```
Comment (cramforce) [inline packages/next/src/utils.ts:2392]: https://www.youtube.com/watch?v=a_B80AIQegE

File: packages/next/src/utils.ts [hunk 2]
```diff
@@ -2352,14 +2352,16 @@ interface MiddlewareManifestV2 {
   functions?: { [page: string]: EdgeFunctionInfoV2 };
 }
 
+type Regions = 'home' | 'edge' | 'auto' | string[] | 'all' | 'default';
```
Comment (EndangeredMassa) [inline packages/next/src/utils.ts:2355]: **question (non-blocking):** ‏

File: packages/next/src/utils.ts [hunk 3]
```diff
@@ -2506,7 +2547,7 @@ export async function getMiddlewareBundle({
                   ...wasmFiles,
                   ...assetFiles,
                 },
-                regions: edgeFunction.regions,
+                regions: normalizeRegions(edgeFunction.regions ?? []),
```
Comment (ijjk) [inline packages/next/src/utils.ts]: Does an empty array behave the same as `undefined`? If so seems good

File: packages/next/src/utils.ts (hunk 2 above)
Comment (EndangeredMassa) [inline packages/next/src/utils.ts]: **question (non-blocking):** ‏Can `'global'` be a value here too or does it only show up in the `string[]`?

File: packages/next/src/utils.ts [hunk 4]
```diff
@@ -2377,6 +2379,45 @@ interface EdgeFunctionMatcher {
...
+}
+
+/**
+ * Normalizes the regions config that comes from the Next.js edge functions manifest.
+ * Ensures that config like `home` and `global` are converted to the corresponding Vercel region config.
+ * In the future we'll want to make `home` and `global` part of the Build Output API.
+ * - `home` refers to the regions set in vercel.json or on the Vercel dashboard project config.
+ * - `global` refers to all regions.
+ */
+function normalizeRegions(regions: Regions): string[] {
```
Comment (EndangeredMassa) [inline packages/next/src/utils.ts:2395]: **suggestion (soft-blocking):** ‏add tests

In this or a follow-up PR: Can we add tests?

**Review by EndangeredMassa — DISMISSED** [review]
I don't have context in the `home`/`global` requirements, but from the code and comments, this makes sense to me.

File: packages/next/src/utils.ts (hunk 3 above)
Comment (timneutkens) [reply packages/next/src/utils.ts]: Changed it to be `undefined` when there are no regions

File: packages/next/src/utils.ts [hunk 5]
```diff
@@ -2377,6 +2379,45 @@ interface EdgeFunctionMatcher {
...
+  const newRegions: string[] = [];
+  for (const region of regions) {
```
Comment (cb1kenobi) [inline packages/next/src/utils.ts:2401]: It appears as if `regions` could be a `string` or `string[]`, so in the event it's a string (e.g. `"home"`), consider arrayifying it:

```suggestion
  if (!Array.isArray(regions)) {
    regions = [regions];
  }

  for (const region of regions) {
```

**Review by cb1kenobi — DISMISSED** [review]

File: packages/next/src/utils.ts (hunk 5 above)
Comment (timneutkens) [reply packages/next/src/utils.ts:2401]: Thank you, super good catch! Got bit by the "string can be iterated" with TypeScript once again 🤦

File: packages/next/src/utils.ts [hunk 6]
```diff
@@ -2377,6 +2379,49 @@ interface EdgeFunctionMatcher {
...
+  }
+
+  const newRegions: string[] = [];
+  for (const region of regions) {
+    // Explicitly mentioned as `home` is one of the explicit values for preferredRegion in Next.js.
+    if (region === 'home' && vercelFunctionRegions) {
+      // Includes the regions from the VERCEL_FUNCTION_REGIONS env var.
+      newRegions.push(...vercelFunctionRegions);
+      continue;
+    }
```
Comment (cb1kenobi) [inline packages/next/src/utils.ts]: If `region === 'home'`, but `vercelFunctionRegions` is falsey, then `"home"` will be pushed into `newRegions` below. Is that to be expected?

If not, then an inner `if` would prevent that:

```suggestion
    if (region === 'home') {
      if (vercelFunctionRegions) {
        // Includes the regions from the VERCEL_FUNCTION_REGIONS env var.
        newRegions.push(...vercelFunctionRegions);
      }
      continue;
    }
```

**Review by ijjk — DISMISSED** [review]

File: packages/next/src/utils.ts (hunk 4 above)
Comment (ijjk) [reply packages/next/src/utils.ts:2395]: We can look at tests in a follow-up yeah

**Review by ijjk — DISMISSED** [review]

**Review by TooTallNate — DISMISSED** [review]

File: packages/next/src/utils.ts (hunk 1 above)
Comment (t3dotgg) [reply packages/next/src/utils.ts:2392]: omfg

**Review by ijjk — APPROVED** [review]

File: packages/next/src/utils.ts (hunk 1 above)
Comment (codercatdev) [reply packages/next/src/utils.ts:2392]: @cramforce nice!

**Pull request merged**